### ISSUE TITLE
Test Training our Own Image Auto-Encoder Based on Manually Processed Image Dataset

### ISSUE THREAD

**Issue by keenanjohnson** [Seattle-Aquarium/CCR_image_processing#8]
We have had limited success evaluating off-the-shelf / pretrained models for our application to automatically process the ROV images.

Previously we evaluated the Sea-Thru method here (https://github.com/Seattle-Aquarium/CCR_image_processing/issues/7), but it's quality was lacking. 

We also investigated many off the shelf ML models (https://docs.google.com/spreadsheets/d/1Dj-OBy53q5kIlxUjKVhsag6B4Bu79B58BEZHhlcIZP0/edit?gid=0#gid=0), but none were trained or optimized for our large format images (most available models were only focused on 256 x 256 pixel images), which is not adequate for our purposes. Thus, we decided we should evaluate training our own model. 

The full unprocessed dataset is available here: https://www.dropbox.com/scl/fo/0jm7jocha7uj5ce9u1vvx/AILmrQqqiT18ghbCHFxDcng?rlkey=jyiiskar6n33miprjcl5qxpaz&st=v51ehwym&dl=0

**Comment by keenanjohnson** [comment]
I have begun the start of a project to make it easy to train our own image auto encoder here: https://github.com/keenanjohnson/underwater-auto-image-encoder.

Currently running the first round of training in google co-lab via the [Jupyter notebook here](https://github.com/keenanjohnson/underwater-auto-image-encoder/blob/main/train_underwater_enhancer_colab.ipynb) which increases the ease of setup. 

Currently training on 512x512 size test images just to validate my setup works. It that is promising, then we can easily increase the size via the parameters in the file. 

<img width="1504" height="848" alt="Image" src="https://github.com/user-attachments/assets/78019296-b439-40bb-a4b2-35de7dd85b6d" />

**Comment by keenanjohnson** [comment]
This model I am training first is based on the [U-shape_Transformer_for_Underwater_Image_Enhancement](https://github.com/LintaoPeng/U-shape_Transformer_for_Underwater_Image_Enhancement) paper as it seems to be the right mix of not too new, and highly cited.

**Comment by keenanjohnson** [comment]
For reference, here were the graphs of the training epochs. 

<img width="1185" height="384" alt="Image" src="https://github.com/user-attachments/assets/df29ca3f-96b6-43eb-b6cc-b80682d75a9d" />

**Comment by zhrandell** [comment]
Regarding the denoise experiements, I downloaded those images and extracted the same patch from all the photos. For example, the hand-edited and the 5000 pair trained model (labeled `cropped_enhance`) patches are below: 

<p float="left">
  <img src="https://github.com/user-attachments/assets/5ff13690-760f-48ca-8e7e-c77ad7bc9757" width="49%" />
  <img src="https://github.com/user-attachments/assets/8d2b0aff-f42b-434a-8b13-70252a0f569c" width="49%" />
</p>

I'll now paste the patches, grouped together as best as possible given their file name: 

<p float="left">
  <img src="https://github.com/user-attachments/assets/01be3397-318d-4adf-982d-2579b5f765da" width="30%" />
  <img src="https://github.com/user-attachments/assets/88148bbf-9072-4663-bece-73c847a926c1" width="30%" />
  <img src="https://github.com/user-attachments/assets/d4004eef-0dc7-49e3-8f9c-0b05db5d6d1e" width="30%" />
</p>

<p float="left">
  <img src="https://github.com/user-attachments/assets/ebd37c49-b282-45ca-a6ce-13f4c2074aed" width="30%" />
  <img src="https://github.com/user-attachments/assets/1db5ccbd-227d-41db-a840-91440159a89f" width="30%" />
  <img src="https://github.com/user-attachments/assets/80f3018a-0918-4517-85fa-7909a16ade9e" width="30%" />
</p>

<p float="left">
  <img src="https://github.com/user-attachments/assets/dcf9bd53-bdb5-4c99-aae2-9b979c532ddb" width="30%" />
  <img src="https://github.com/user-attachments/assets/590d0a2e-70fe-451d-b5f1-af292816cec2" width="30%" />
</p>

<p float="left">
  <img src="https://github.com/user-attachments/assets/6b1470ae-dff3-4c08-ba3b-1735f8549c2e" width="30%" />
  <img src="https://github.com/user-attachments/assets/1f5bba39-018f-4566-a0cb-bc83eea474f6" width="30%" />
</p>

It's clear to see the tradeoff you mentioned, @keenanjohnson, between image clarity and noise. It seems that just about any steps to address the noise reduces the clarity. 

I couldn't help but come away with a new appreciation for Adobe Lightroom's Denoise Feature, as it appears (based on examining the hand-edited) to reduce noise while sharpening edges (i.e., without a reduction in clarity). In reading a bit online, while the specifics of the algorithm don't appear to be listed, what I can see is that Lightroom's Denoise works best when applied to the raw photos, as it jointly demosaics + denoises in a single ML pass. It appears to be a ML solution that's trained on millions of high-noise / low-noise raw patch pairs, so it learns to scrub fixed-pattern/pixel-level shadow junk without smearing detail. [This Adobe blog post](https://blog.adobe.com/en/publish/2023/04/18/denoise-demystified) provided useful info.

I tried to identify some open-source solutions that were as close to this as possible, and one option appears to be [DarkTable](https://www.darktable.org/). While it is open-source, it appears to be more of a GUI, though I think it has some API functionality. It looks like it has a lot of [neat tools](https://docs.darktable.org/usermanual/3.8/en/module-reference/processing-modules/denoise-profiled/) to work with raw photos, though when I dug in further, I see that GoPro photos are [not supported](https://www.darktable.org/resources/camera-support/) (for camera-specific tools, which it appears the denoise is). 

I think one additional test that would be useful is for @reidtosa to run Lightroom's Denoise Feature on our 20 raw test images, and then apply the 5000 pair model to those photos. It may end up being the case that the Denoise Feature is something we want to retain from Lightroom (which wouldn't be onerous, as it requires no hand-tuning).

**Comment by keenanjohnson** [comment]
Hey all! As i mentioned above, I continued to let the model train for a few more iterations in co-lab. The initial 5000 image model I shared ran for 50 training interations. This one I let run until 80 iterations. Each iteration on google colab took around 3 hours, so this represents around 240 hours of compute time to train the model using our current method to this point. That model is here: https://drive.google.com/drive/u/0/folders/11bQRjNoMbC_yTFNzPvpZkyb1INf61ajf

The new model has these specs:
```
Train Loss: 0.0212, Train PSNR: 28.80 dB
Val Loss: 0.0218, Val PSNR: 28.51 dB
```

Compared to our previous 50 iteration model at:

```
Train Loss: 0.0234, Train PSNR: 28.03 dB
Val Loss: 0.0245, Val PSNR: 27.76 dB
```

The different isn't large, around .8 dB gain and less than half a percent improvement, so it was honestly hard for me to spot much different reviewing images, but curious to see what you all think in the comparison now that you can run the models on your own!

Additionally, I was curious if you had tried processing any images in coralnet toolbox with these ML preprocessed images and how the results look?

My key takeaways from the effort so far:

1. Training an ML model to preprocess the images does seem quite feasible, we just need to arrive on the right method and training set
2. Training large models like this in google colab is possible, but google colab kept disconnecting and generally being somewhat unreliable the longer it went, so given the hundreds of hours it takes to train these, we need a different solution (via a cloud vm or otherwise)
3. I feel like we are approaching the limit of quality with the current u-net model, so it might be worth trying to try some other models from the papers that we researched previous for you all to evaluate as a next step. 
4. Denoising through the machine learning model alone is challenging and using more algorithmic methods has also not proved to have great results so far. It's unclear if denoising is needed, but investigating the best techniques there that don't involve light room is still open.
5. We will want to probably continuously train new models as the scope of our surveys expand and new features / areas are added to our available data to make our preprocessing more useful broadly.

So in terms of actionable tasks:

- Evaluate the current setup results with the coralnet-toolbox
- Set up a training system that is easy to start, but doesn't involved google colab so we can kick off training runs without fear of it stopping and having to check on it often (Zach I'll probably start this on the VM that you mentioned goes away at the end of sep.)
- Investigate new denoising methods
- Identify the next most promising machine learning method from the literatures, train it with our dataset and compare to unet etc.

Does that all make sense to everyone, am I missing something? If so, I can create sub issues for each of those things above.

**Comment by zhrandell** [comment]
@keenanjohnson, I think you hit the nail on the head with all of these takeaways and actionable items - thank you! And likewise, thank you for maintaining the training for 240 hrs . . . it indeed sounds like you've squeezed as much out of these training image pairs and the u-net framework as possible. 

A few thoughts in response to your takeaways and actionable tasks:  

- For multiple reasons, I agree that it makes sense to have the model training take place in a VM environment such as the Google TPU cloud. If needed, we have grant $ we can invoke to add more compute / extend cloud access, and we could likely apply to Google to maintain free or low-cost access to compute resources once we have something cool up and running; let's do it! 

- Speaking of model training, I did apply the UIE (using the previous model trained on all 5000 images) to GPR imagery of uchin barrens from offshore of Neah Bay - the visual appearance of these urchin barren photos is very distinct from anything the UIE had seen before in its training imagery; unsurprisingly, the UIE output wasn't great. So, this is evidence that we will indeed want the ability to easily train new models for new areas of ROV operations, etc. 

- Re: de-noise (and the ability for us to eventually train models ourselves): right now on the Aquarium's mapping station computer we're running Adobe Lightroom's de-noise filter on the full set of ~ 5,100 raw GPRs (it's going to take a few days!). As a future experiment, we could train a new auto-encoder model using those denoised images as the input. The resulting model might have output closer to our original, hand-edited Adobe Lightroom processed imagery. 

-   Expanding upon the above, however, I view that structured (and hypothetically, perpetual) future usage of Adobe Lightroom's denoise in the model training process as a "fall back" option in the event that open-source denoise filters prove unreliable, AND that testing reveals the importance of denoise when it comes to extracting data from Toolbox. 

- Speaking of extracting data from Toolbox, this upcoming week we will try to conduct the following tests: (**1**) apply our existing, trained Toolbox model to the (somewhat) noisy UIE output data and evaluate model prediction % accuracy. As a more robust and complete test, we will also (**2**) train a new Toolbox model using the exact same image patch locations, but with UIE processed imagery containing noise - we'll evaluate the confusion matrix and the test accuracy from the train/validation/testing phases of model compilation. This latter test in particular will shed a lot of light on just how important or not de-noise is re: Toolbox efficacy (as we can directly compare it to the current model test accuracy). These tests will be tracked at https://github.com/Seattle-Aquarium/CCR_image_processing/issues/11

- Finally, I fully support exploring other models beyond the current u-net framework . . . @tonioteran, do you have any thoughts on other compelling modeling frameworks that would be interesting to test?? We still have the Google sheet linked [here](https://docs.google.com/spreadsheets/d/1Dj-OBy53q5kIlxUjKVhsag6B4Bu79B58BEZHhlcIZP0/edit?usp=sharing) in https://github.com/Seattle-Aquarium/CCR_image_processing/issues/3

**Comment by clydemcqueen** [comment]
Are the file names in input_TIFF off-by-one? E.g., it looks like these are all related:
* input_GPR/2024_09_10_11-58-35.GPR
* input_TIFF/2024_09_10_11-58-34.tif
* human_output_JPEG/2024_09_10_11-58-35.jpg

**Comment by reidtosa** [comment]
Good catch, @clydemcqueen. I believe this is a strange Lightroom error we've encountered once before. Not sure as to the cause however since the TIF files should be taking the GPR names exactly. I'll correct that ASAP.

**Comment by zhrandell** [comment]
I'm going to close this for now . . . the preponderance of evidence suggests that using our hand-edited imagery to train an auto-encoder is absolutely a viable area of development, and we plan to fully proceed. We're now working through the details (e.g., U-net vs U-shape model architecture, as detailed here: https://github.com/Seattle-Aquarium/CCR_image_processing/issues/15, and how we partition the training imagery, as discussed here: https://github.com/Seattle-Aquarium/CCR_image_processing/issues/14)
